tracing/config.go

```diff
@@ -1,23 +1,55 @@
 package tracing
 
+import (
+	"bytes"
+	"embed"
+	"io"
+)
+
 // JaegerConfig encapsulates jaeger's configuration.
 type JaegerConfig struct {
-	LocalAgentHostPort string
-	SamplerType        string
-	SamplerValue       float64
-	SamplerServerURL   string
-	Propagation        string
-	MaxTagValueLength  int
+	LocalAgentAddress string          `json:"local_agent_address"`
+	Sampling          *JaegerSampling `json:"sampling"`
+	Propagation       string          `json:"propagation"`
+	MaxTagValueLength int             `json:"max_tag_value_length"`
+}
+
+type JaegerSampling struct {
+	Type      string  `json:"type"`
+	Value     float64 `json:"value"`
+	ServerURL string  `json:"server_url"`
 }
 
 // ZipkinConfig encapsulates zipkin's configuration.
 type ZipkinConfig struct {
-	ServerURL string
+	ServerURL string `json:"server_url"`
 }
 
 type Config struct {
-	ServiceName string
-	Provider    string
-	Jaeger      *JaegerConfig
-	Zipkin      *ZipkinConfig
+	ServiceName string           `json:"service_name"`
+	Provider    string           `json:"provider"`
+	Providers   *ProvidersConfig `json:"providers"`
+}
+
+type ProvidersConfig struct {
+	Jaeger *JaegerConfig `json:"jaeger"`
+	Zipkin *ZipkinConfig `json:"zipkin"`
+}
+
+//go:embed config.schema.json
+var ConfigSchema embed.FS
+
+const ConfigSchemaID = "https://ory.sh/schemas/tracing.schema.json"
+
+// AddConfigSchema adds the tracing schema to the compiler.
+// The interface is specified instead of `jsonschema.Compiler` to allow the use of any jsonschema library fork or version.
+func AddConfigSchema(c interface {
+	AddResource(url string, r io.Reader) error
+}) error {
+	schema, err := ConfigSchema.ReadFile("config.schema.json")
+	if err != nil {
+		panic("unrecoverable error: could not read file from embed.FS: " + err.Error())
+	}
+
+	return c.AddResource(ConfigSchemaID, bytes.NewBuffer(schema))
```

Comment:
> 1. You are referencing a non-versioned URL. This will eventually update, be refactored, changed. All older builds will no longer work as soon as this changes. If you want to reference a URL, it must be versioned (like a git hash).
> 2. You are referencing a HTTP URL. This means that Ory Keto will not work in envs without network access. We got hit by this in Ory Oathkeeper as well

The URL is just a random placeholder., it only has to be unique The schema is loaded from the `embed.FS`, versioned by the version of ory/x you have in the `go.mod`. That is why we have to add it to the compiler here @aeneasr 

We can instead use e.g. the string `ory/schemas/tracing` or just anything we can assume to be unique.

Reply:
yeah, a fixed string will probably prevent the loader from trying to load it from remote if it's not available for some reason (e.g. build error) :)